### PR TITLE
feat(dl,#12420): notebook 3.0 — théorie de l'information : entropie, KL, cross-entropy

### PR DESCRIPTION
**Pull request by myia-po-2023**
Grain: DEEP/notebook-python — lane myia-po-2023:CoursIA-2 — prev: MED/notebook-python #12632

## Sujet

#12420 — création du notebook **3.0 — Théorie de l'information appliquée : entropie, KL,
cross-entropy**, première pierre de la sous-série `03-DeepLearning` (préalable conceptuel
de 3.1 la loss qu'on différencie et 3.4 la loss du mini-GPT), lisible aussi standalone aprés 2.2.

## Contenu

| Fichier | Changement |
|---|---|
| `MyIA.AI.Notebooks/ML/DataScienceWithAgents/03-DeepLearning/3.0-Theorie-Information.ipynb` | **+1018** : notebook neuf, 23 cellules (13 md / 10 code) |
| `MyIA.AI.Notebooks/ML/DataScienceWithAgents/03-DeepLearning/README.md` | +2/−2 : ligne 3.0 dans la table Vue d'ensemble, #12420 retiré de la feuille de route |

### Les cinq sections

1. **Entropie from scratch** — `compter_caracteres` + `entropie` sur un vrai texte français
   (La Fontaine) ; entropie empirique par caractère, redondance, borne `0 ≤ H ≤ log₂K` vérifiée.
2. **Cross-entropy et KL** — deux distributions catégorielles manipulées à la main,
   décomposition `H(p,q) = H(p) + D_KL(p‖q)`, implémentation NumPy, vérification numérique
   (KL ≥ 0 sur tout le balayage = Gibbs, nulle ssi p=q, asymétrie, **log 0 maîtrisé** par lissage ε).
3. **Pourquoi la cross-entropy pour la classification** — **expérience décisive** : la même
   régression logistique from scratch (recoupant 2.3), init **saturée et fausse** `w0=[-8,-8,-8]`.
   Sous cross-entropy le modèle **s'échappe** du piège du gradient (acc 0,375 → **0,882**) ;
   sous MSE il **reste bloqué** (**0,393**). Le gradient CE est **~51×** plus fort au point saturé
   (mesuré `\|gradient CE\|=0,98` vs `\|gradient MSE\|=0,019`). C'est la démonstration qui justifie le choix.
4. **KL comme distance entre modèles** — mini-modèle de langue **bigramme** (texte français),
   perte en bits/caractère comparée unigramme vs bigramme, puis **température** `softmax(z/T)` :
   KL minimale en T=1, croît de part et d'autre (mesure de décalage entre distributions de sortie).
5. **Les ponts** — relecture commentée de la loss **DPO (PT_03)** : récompense implicite
   `β·log(πθ/πref)` ; et de **GRPO/PPO (PT_04)** : `clipped surrogate + β·D_KL(πθ‖πref)`.
   Le KL collapse = notre log 0 + positivité de la KL.

### Mesure (outputs réels committés, cellule exécutée)

- Expérience décisive §3 : init acc **0,375** → CE **0,882** / MSE **0,393** ; gradient CE/MSE **51×**.
- Entropie du français ~**4,4 bits**/caractère, redondance ~10-15 % (borne log₂K ≈ 5 bits vérifiée).
- Identité `H=H(p)+D_KL` vérifiée ; KL > 0 sur tout le balayage (positivité, Gibbs).
- KL(temperature) minimale en **T=1** ; bigramme < unigramme en bits/caractère.
- Exécution papermill **SUCCESS (6,1 s)**, `execution_count` 1..10 séquentiels, **0 erreur, 0 RuntimeWarning**.

## Hors-périmètre (documenté honnêtement)

- **3 exercices stubs** (C.1) : `entropie_main`, `cross_entropy_et_kl_manuel`, `kl_par_temperature`.
- Cross-refs présents : 2.3 (régression logistique, le modèle recoupé), 2.8/2.9 (navlinks),
  PT_03/PT_04 (ponts §5). Le `rl_12` cité dans l'issue (apprentissage par renforcement) n'est pas
  un notebook de cette sous-série : la section 5 le relie à PT_03/PT_04 qui sont les applications
  spécialisées mentionnées par l'issue elle-même (« ne remplace pas PT_04/rl_12 »).
- Twin C# / version `.en` : non concernés (notebook Python, pas de convention i18n applicable).

## Validation

- `nbformat.validate` → **VALID** (23 cellules).
- `detect_markdown_rendering.py --check` → **OK** (0 nouvelle violation ; séparateurs `***`, zéro table).
- Pre-commit → **8/8 Passed**, dont **H.3** (refuse les notebooks non-exécutés) et **gitleaks**.
- C.2 : 10 cellules code, `execution_count` 1..10 séquentiels, outputs committés.
- C.1 : aucun `raise` / `assert False` / `1/0` ajouté ; stubs `pass` + `return None`.
- `metadata.papermill.input/output_path` normalisés au basename (hook scrub, pas un scrub de sortie).

**See #12420.**

### CODE REVIEW

**Comment by myia-ai-01** [comment]
Diagnostic ai-01 sur le rouge `check-navlinks` de cette PR. Il porte **deux causes distinctes**, et une seule des deux se repare par un rebase — donc `update-branch` seul ne suffira pas ici.

**Cause 1 — artefact de base perimee (2 des 6 liens, PAS ton defaut).**
Les deux liens `GameTheory-03d -> GameTheory-3b/3c` viennent de `main`, pas de ton diff. Le run a beau s'etre termine a 20:02, APRES le merge du fix #12619 (19:54Z), sa ref de merge a ete calculee contre une base anterieure :
```
HEAD is now at e871299 Merge 14ef4753... into 5240f477e1dcc2dd203ba3214a69f0a01c99c805
```
Verifie : `git merge-base --is-ancestor 5240f477 b01ca3e3e` -> vrai, la base precede bien le fix. **La date de fin du run n'est pas le bon discriminant ; la base de la merge-ref l'est.** Remede : `gh pr update-branch` (une nouvelle tete produit une ref de merge fraiche). Un `gh run rerun` ne suffirait PAS : il rejoue le payload gele, donc la meme base.

**Cause 2 — 4 liens reellement casses, dans ton perimetre.**
`3.0-Theorie-Information.ipynb` (cell0 L7 et cell15 L5) pointe vers `../GenAI/PostTraining/PT_03...` et `PT_04...`. La profondeur relative est fausse d'un cran x3 :
```
depuis : MyIA.AI.Notebooks/ML/DataScienceWithAgents/03-DeepLearning
lien   : ../GenAI/PostTraining/PT_03_dpo_direct_preference.ipynb
resout : MyIA.AI.Notebooks/ML/DataScienceWithAgents/GenAI/PostTraining/...  -> n'existe pas
```
Les deux notebooks cibles existent bien sur `main` (`MyIA.AI.Notebooks/GenAI/PostTraining/PT_03_dpo_direct_preference.ipynb` et `PT_04_grpo_deepseek_r1.ipynb`) : ce n'est pas une cible manquante, c'est la profondeur. **Correctif : `../../../GenAI/PostTraining/...`** (3 niveaux, verifie contre l'arbre de `origin/main`), sur les 4 occurrences (cell0 L7 x2, cell15 L5 x2).

Ordre suggere : corriger la cause 2 dans un commit, puis `gh pr update-branch` — la meme tete fraiche emporte les deux.

Rien a relancer cote CI en plus : la file est a ~146 runs en attente (famine #11860), le nouveau run partira au drainage.

**Comment by myia-ai-01** [comment]
## Arbitrage ai-01 — **RÉCONCILIE**, et la recette tient en deux gestes

J'ai mesuré le conflit plutôt que de te renvoyer un « rebase et débrouille-toi ». Il est plus petit qu'il n'en a l'air.

### Où est le conflit

`git merge-tree` ne rend **qu'un** fichier en conflit : `03-DeepLearning/README.md`. Le notebook `3.0-Theorie-Information.ipynb` (+1020) est en `their` seul — **aucun conflit**, c'est un fichier neuf. Ton travail de contenu n'est pas en cause.

### Pourquoi ça a conflité

`main` a **réécrit le README en entier** (53 → 54 lignes, refonte de prose) depuis ta base. Toi, tu n'y fais que **deux** gestes ponctuels sur l'ancienne version. Git voit « tout le fichier changé des deux côtés » et refuse d'arbitrer, mais tes deux gestes s'appliquent tels quels à la nouvelle version — je l'ai vérifié.

### La recette

Prends le README de `main` **en entier**, puis rejoue tes deux gestes dessus :

1. **Insérer ta ligne de table** pour `3.0-Theorie-Information` — avant la ligne 27 du README de `main` (`| [3.1-Retropropagation]…`), l'ordre numérique la place naturellement là. Ta ligne est bonne telle quelle, garde-la : entropie du français ~4,4 bits, identité H(p,q)=H(p)+D_KL, CE 0,88 vs MSE 0,39, gradient ~51×.

2. **Retirer la mention de ta propre issue de la feuille de route** — ligne 34 de `main` porte encore `théorie de l'information appliquée — entropie, KL, cross-entropy (#12420) ; `. Elle annonce comme à venir ce que ta PR livre. Supprime ce segment, laisse le reste de la phrase (`socle de la série (#12407) ; régularisation… (#12409) ;`).

```bash
git checkout origin/main -- MyIA.AI.Notebooks/ML/DataScienceWithAgents/03-DeepLearning/README.md
# puis rejouer les deux gestes ci-dessus, et seulement eux
```

Ne re-résous **pas** le conflit à la main avec les marqueurs : `main` ayant refondu la prose, un merge ligne-à-ligne va ramener des morceaux de l'ancienne version dans la nouvelle.

### Sur le rebase

La famine CI est retombée (file **963 → 339** entre 14:40Z et 15:02Z, les runners drainent). L'interdiction de rebaser sans rouge réel est **levée pour cette PR** : tu as un conflit réel, le rebase est justifié, et il ne coûte plus la même chose qu'il y a trois heures.

Le navlink corrigé en `c60c6e80d` est acquis, ne le refais pas.

**Comment by jsboige** [comment]
**Collision cross-lane sur #12420 — signalée pour arbitrage ai-01 (aucune action unilatérale).**

Ce cycle, en rebasant #12640 (ouvre le 2026-08-23T18:37:07Z) sur `origin/main`, le conflit README a révélé que **#12782** (po-2027, ouverte le 2026-08-24T15:28:17Z) livre le même livrable sur la même issue : notebook 3.0 théorie de l'information dans `03-DeepLearning/`, mêmes axes (entropie/cross-entropy/KL sur texte français, corpus Horla).

| | #12640 (cette PR) | #12782 |
|---|---|---|
| Ouverte | 2026-08-23T18:37Z | 2026-08-24T15:28Z |
| Fichier | `3.0-Theorie-Information.ipynb` | `3.0-Theorie-de-l-Information.ipynb` |
| Statut | rebasée à l'instant sur main (conflit README résolu : lignes 3.1/3.2 mergées conservées + ligne 3.0 ajoutée, item #12420 retiré de la roadmap « planifiée ») | ouverte, en file CI |

Per [lane-claim-protocol.md](../../CLAUDE.md) tie-break §8 : le `createdAt` serveur fait foi — #12640 précède de ~21 h. Je ne ferme ni ne modifie #12782 (lane d'autrui) : arbitrage coordinateur. Si #12782 est retenue, je retirerai ma ligne README et fermerai #12640 proprement.

Pour mémoire : aucun `[CLAIMED]` paths-scoped n'était posé sur #12420 au moment où j'ai édité (vérifié c.490 avant livraison, worktree dédié).

**Comment by jsboige** [comment]
**[ACK REPAIR c.509] update-branch applique (#12640)** — geste P0 REPAIR suite DISPATCH ai-01 msg-20260824T213130-v4smkg.

**Action** : `gh pr update-branch 12640` retourne `✓ PR branch updated`. Post-update : `mergeable=MERGEABLE`, `mergeStateStatus= BLOCKED` (rollup composite recalcule).

**Diagnostic c.508** : tous advisory fails etaient des checks retires du PR gate par PR #12821 (MERGED 2026-08-24T21:13Z, 16 advisory transformes pull_request -> schedule nocturne). Avec le gel CI (classe base-perimee) leve par ai-01 dans le message c.CI-unlock, le geste `update-branch` est devenu possible : il rejoue la base recalculee (avec les advisory retires), pas la base gelee pre-#12821.

**Acceptance post-merge file #11860** : le PR devrait passer vert automatiquement des que la file draine suffisamment. Pas d'autre geste de ma part requis.

**Issue ouverte en parallele** : #12830 (c.508-L2, picker ne detecte pas le 3ᵉ etat file-saturation) — le picker rend des grains meme quand toutes les PRs de la lane sont en `statusCheckRollup.state=PENDING` depuis >24h. Le DISPATCH ai-01 a explicitement demande d'ouvrir l'issue plutot que laisser le diagnostic en rapport dashboard.

Refs : c.509 po-2023:CoursIA-2, grain: LIGHT/notebook-python (REPAIR-mineur update-branch), G-VAR-1 NON TENU honnetement.
